inc/sainat.h

```diff
@@ -0,0 +1,653 @@
+/**
+ * Copyright (c) 2014 Microsoft Open Technologies, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *    not use this file except in compliance with the License. You may obtain
+ *    a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
+ *    CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
+ *    LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
+ *    FOR A PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
+ *
+ *    See the Apache Version 2.0 License for specific language governing
+ *    permissions and limitations under the License.
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc., Marvell International Ltd.
+ *
+ * @file    sainat.h
+ *
+ * @brief   This module defines SAI NAT (Network Address Translation) spec
+ */
+
+#if !defined (__SAINAT_H_)
+#define __SAINAT_H_
+
+#include <saitypes.h>
+
+/**
+ * @defgroup SAINAT SAI - Network Address Translation (NAT) specific API definitions
+ *
+ * @{
+ */
+
+/**
+ * @brief NAT Type
+ */
+typedef enum _sai_nat_type_t
+{
+    /** No NAT */
+    SAI_NAT_TYPE_NONE,
+
+    /** Source NAT */
+    SAI_NAT_TYPE_SOURCE_NAT,
+
+    /** Destination NAT */
+    SAI_NAT_TYPE_DESTINATION_NAT,
+
+    /** Double NAT */
+    SAI_NAT_TYPE_DOUBLE_NAT,
+
+} sai_nat_type_t;
+
+/**
+ * @brief NAT Entry Attributes for Match
+ */
+typedef enum _sai_nat_entry_attr_t
+{
+
+    /**
+     * @brief Start of Attributes
+     */
+    SAI_NAT_ENTRY_ATTR_START,
+
+    /**
+     * @brief NAT Type defined in sai_nat_type_t
+     * @type sai_nat_type_t
+     * @flags CREATE_AND_SET
+     * @default SAI_NAT_TYPE_NONE
+     */
+    SAI_NAT_ENTRY_ATTR_NAT_TYPE  = SAI_NAT_ENTRY_ATTR_START,
```

Comment:
If we are not able to accommodate Marian's proposal of different APIs, the next best thing would be to move the SAI_NAT_ENTRY_ATTR_NAT_TYPE to be part of sai_nat_entry_key_t.
This allows for easier documentation.

typedef struct _sai_nat_entry_key_t
	{
	sai_nat_type_t nat_type;
	    /**
	     * @brief IPv4 source address
>> valid only if nat_type is DoubleNAT or SNAT
	     */
	    sai_ip4_t src_ip;
	
	    /**
	     * @brief IPv4 destination address
>> valid only if nat_type is DNAT or DoubleNAT
	     */
	    sai_ip4_t dst_ip;
	:
        :

Further, SAI_NAT_ENTRY_ATTR_NAT_TYPE  is marked as CREATE_AND_SET -- modifying this at run-time actually means moving the entry from one logical block to another. It looks better to mark it as create_only.

Reply:
Ravi, 
As we discussed in the community meeting, moving NAT type to NAT entry key is correct from the pipeline perspective. NAT type is an attribute which is a result of the match.

Regarding CREATE_AND_SET, thats a good suggestion. I will change it CREATE_ONLY. This does not look for a good candidate for add/modify so the proposed change makes sense

Reply:
Ravi,

I just realized why I marked this as CREATE_AND_SET. metadata checker allows validly check only with flags CREATE_AND_SET and not CREATE_ONLY. Sorry I have revert back this change